### ISSUE TITLE
Bit error rate simulator corrupts PCAP savefile format

### ISSUE THREAD

**Issue by digital-nomad8620** [oresat/oresat-libdxwifi#27]
Whenever offline packet transmission is run with the `--error-rate` flag set, the resulting `savefile` is unreadable by the `rx` program. My best guess is that somehow the per-file and per-packet headers are being corrupted when we randomly flip bits. 

See the savefile format here: https://www.tcpdump.org/manpages/pcap-savefile.5.html

Here's a sample run and resulting output: 

"transmitting" the file
```
./bin/TestDebug/tx --savefile daisy.tx daisy.enc -e 0.2                                                                                                                                  
[ INFO ][ transmitter ] : DxWifi Transmitter Settings
        Device:              mon0
        PA Enabled:          0
        Block Size:          1275
        Transmit Timeout:    -1
        Redundant Ctrl:      0
        Data Rate:           1Mbps
        RTAP flags:          0x10
        RTAP Tx flags:       0x8

[ INFO ][ tx ] : Opened daisy.enc for transmission
[ INFO ][ transmitter ] : Starting DxWiFi Transmission...
[ INFO ][ transmitter ] : DxWiFI Transmission stopped
[ INFO ][ transmitter ] : DxWifi transmitter closed
```

"receiving" the file
```
./bin/TestDebug/rx --savefile daisy.tx daisy.rx -v                                                                                                                                           0.06    INSERT
[ INFO ][ receiver ] : DxWifi Receiver Settings
        Device:                   mon0
        Capture Timeout:          -1s
        Packet Buffer Size:       5242880
        Ordered:                  0
        Add-noise:                0
        Filter:                   wlan addr2 aa:aa:aa:aa:aa:aa
        Optimize:                 1
        Snapshot Length:          65535
        PCAP Buffer Timeout:      20ms
        Dispatch Count:           1
        Datalink Type:            802.11 plus radiotap header

[ INFO ][ receiver ] : Starting packet capture...
[ FATAL ][ receiver ] : status: 0
[ INFO ][ receiver ] : DxWiFi Reciever capture ended
[ WARN ][ receiver ] : Failed to gather capture stats from PCAP
[ DEBUG ][ rx ] : Receiver Capture Stats
        Total Payload Size:          0
        Total Write length:          0
        Total Capture Size:          0
        Total Blocks Lost:           0
        Total Noise Added:           0
        Packets Processed:           0
        Packets Received:            0
        Packets Dropped (Kernel):    0
        Packets Dropped (NIC):       0
        Note: Packet drop data is platform dependent.
        Blocks lost is only valid when `ordered` flag is set

[ INFO ][ receiver ] : DxWiFi receiver closed
```

**Comment by digital-nomad8620** [comment]
Also, when a fix for this is done, we should add a test case for this scenario in the automated tests run in `test_tx_rx.py`. The reason this hasn't been caught before is because the current test suite does not exercise the `bit error rate` and `packet loss` features.

**Comment by AlexanderJDupree** [comment]
@khanu263 , @CD-W

**Comment by digital-nomad8620** [comment]
@khanu263 @CD-W 

@Cam57 and I just spent the last hour hex dumping files and pouring through the source code and we found the issue! Good news is that we fixed part of the problem! Bad news is that the problem is a lot worse than we thought.

Here's our findings:

First, my initial hunch that the PCAP savefile format being corrupted was dead wrong. But we were corrupting the [radiotap headers](https://www.radiotap.org/) which was causing PCAP to not recognize the capture format on the receiver. This is something that cannot happen in a production environment since the Radiotap header is discarded before injection and a new Radiotap header is attached whenever the driver receives a packet. So the Radiotap header will never be corrupted in any practical scenario. We fixed this issue in [fbf42b0](https://github.com/oresat/oresat-dxwifi-software/commit/fbf42b090223a8bc3d132098b446a570784a2532) and the receiver was now able to read the savefiles. . . .sometimes. 

The real underlying issue is that the filter we use with PCAP to identify which packets came from the satellite is too strict. The default filter we use is this: 

```
wlan addr2 aa:aa:aa:aa:aa:aa
```

Which basically says "Capture everything over wireless lan whose second address field is aa:aa:aa:aa:aa". The problem arises when any of the bits in the address field is flipped then PCAP will just ignore that packet. Unfortunately, this is a fairly common error. When @Cam57 were running tests we were only able to capture 2/11 packets with an error rate of 0.1%. Which is obviously abysmal. 

*Solutions?*

Probably the most straight forward solution is to use a more relaxed filter. However, there is a fine line between too strict and too relaxed. If we relax the filter too much then we will start capturing anyone's wireless traffic in the vicinity which could cause problems for FEC down the line. 

Bottom line, we need to come up with a [BPF](https://en.wikipedia.org/wiki/Berkeley_Packet_Filter) that's relaxed enough for PCAP to capture the packet, but unique enough that random traffic won't get mixed in with those of the satellite. We may also want to program another filter that we apply after we receive the packet that makes sure the packet matches certain criteria (like capture length) before we process it.

**Comment by digital-nomad8620** [comment]
@khanu263, @Cam57 and I are going to spend the weekend experimenting with different filters and then report our findings on Monday. As far as this issue is concerned, I'll open a PR for the Radiotap header corruption issue and then open a new issue for the problem I stated above.

**Comment by khanu263** [comment]
OK, good find! I'm glad the radiotap header got fixed, and I'm _very_ glad we uncovered this address issue. The plan to experiment over the weekend sounds good, and we should indeed move this to a new issue, so thanks for that.
